File: internal/invoice/invoice.go
// Package invoice implements the payment lifecycle: creating invoices, locking a
// receive target, and advancing each invoice's status from on-chain
// observations. It is chain-agnostic — all coin specifics live behind the
// chain.Adapter interface.
package invoice

import (
	"time"

	"github.com/vexarnetwork/vexpay/internal/chain"
	"github.com/vexarnetwork/vexpay/internal/money"
)

// Status is the lifecycle state of an invoice.
type Status string

const (
	// StatusPending: created, awaiting payment.
	StatusPending Status = "pending"
	// StatusUnderpaid: a payment arrived but for less than requested.
	StatusUnderpaid Status = "underpaid"
	// StatusConfirming: sufficient funds seen, waiting for confirmations.
	StatusConfirming Status = "confirming"
	// StatusPaid: fully paid and confirmed.
	StatusPaid Status = "paid"
	// StatusOverpaid: confirmed, but more than requested was sent.
	StatusOverpaid Status = "overpaid"
	// StatusExpired: the payment window closed before settlement.
	StatusExpired Status = "expired"
)

// IsTerminal reports whether a status will never change again.
func (s Status) IsTerminal() bool {
	switch s {
	case StatusPaid, StatusOverpaid, StatusExpired:
		return true
	default:
		return false
	}
}

// IsSettled reports whether the invoice can be considered successfully paid.
func (s Status) IsSettled() bool {
	return s == StatusPaid || s == StatusOverpaid
}

// Invoice is a single request for payment.
type Invoice struct {
	ID         string
	MerchantID string

	Chain chain.ID
	Asset money.Asset

	// Amount is the crypto amount requested.
	Amount money.Amount

	// Optional fiat pricing captured at creation (rate locked for the window).
	FiatCurrency string
	FiatAmount   string
	Rate         string

	// Receive target details.
	ReceiveAddress  string
	PaymentURI      string
	Strategy        chain.ReceiveStrategy
	DerivationIndex uint32

	Status        Status
	Received      money.Amount
	Confirmations int
	RequiredConfs int
	TxHash        string

	Metadata map[string]string

	CreatedAt time.Time
	ExpiresAt time.Time
	PaidAt    *time.Time
	UpdatedAt time.Time
}

// StatusChange describes a transition, emitted to subscribers (e.g. webhooks).
type StatusChange struct {
	Invoice  *Invoice
	Previous Status
}

File: sdks/php/src/Invoice.php
<?php

declare(strict_types=1);

namespace VexPay;

/**
 * An immutable view of a VexPay invoice.
 *
 * Monetary values (amount, received, rate) are exact decimal strings — never
 * cast them to float for comparisons or accounting.
 */
final class Invoice
{
    /**
     * @param array<string,string> $metadata
     * @param array<string,mixed>  $raw       The full decoded API payload.
     */
    public function __construct(
        public readonly string $id,
        public readonly string $merchantId,
        public readonly string $chain,
        public readonly string $asset,
        public readonly string $amount,
        public readonly string $status,
        public readonly string $received,
        public readonly int $confirmations,
        public readonly int $requiredConfirmations,
        public readonly string $receiveAddress,
        public readonly string $paymentUri,
        public readonly string $txHash,
        public readonly string $fiatCurrency,
        public readonly string $fiatAmount,
        public readonly string $rate,
        public readonly array $metadata,
        public readonly ?string $createdAt,
        public readonly ?string $expiresAt,
        public readonly ?string $paidAt,
        public readonly ?string $updatedAt,
        public readonly array $raw,
    ) {
    }

    /**
     * Build an Invoice from a decoded API response array.
     *
     * @param array<string,mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: (string) ($data['id'] ?? ''),
            merchantId: (string) ($data['merchant_id'] ?? ''),
            chain: (string) ($data['chain'] ?? ''),
            asset: (string) ($data['asset'] ?? ''),
            amount: (string) ($data['amount'] ?? '0'),
            status: (string) ($data['status'] ?? ''),
            received: (string) ($data['received'] ?? '0'),
            confirmations: (int) ($data['confirmations'] ?? 0),
            requiredConfirmations: (int) ($data['required_confirmations'] ?? 0),
            receiveAddress: (string) ($data['receive_address'] ?? ''),
            paymentUri: (string) ($data['payment_uri'] ?? ''),
            txHash: (string) ($data['tx_hash'] ?? ''),
            fiatCurrency: (string) ($data['fiat_currency'] ?? ''),
            fiatAmount: (string) ($data['fiat_amount'] ?? ''),
            rate: (string) ($data['rate'] ?? ''),
            metadata: array_map('strval', $data['metadata'] ?? []),
            createdAt: isset($data['created_at']) ? (string) $data['created_at'] : null,
            expiresAt: isset($data['expires_at']) ? (string) $data['expires_at'] : null,
            paidAt: isset($data['paid_at']) ? (string) $data['paid_at'] : null,
            updatedAt: isset($data['updated_at']) ? (string) $data['updated_at'] : null,
            raw: $data,
        );
    }

    /** True once the invoice is fully paid (paid or overpaid). */
    public function isPaid(): bool
    {
        return $this->status === 'paid' || $this->status === 'overpaid';
    }

    /** True once the invoice can no longer change (paid, overpaid or expired). */
    public function isTerminal(): bool
    {
        return $this->isPaid() || $this->status === 'expired';
    }
}
